apollo-core/src/main/java/com/ctrip/framework/apollo/core/enums/Env.java

```diff
@@ -17,13 +24,142 @@
  * </ul>
  *
  * @author Jason Song(song_s@ctrip.com)
+ * @changeby wxq
  */
-public enum Env{
-  LOCAL, DEV, FWS, FAT, UAT, LPT, PRO, TOOLS, UNKNOWN;
+public final class  Env {
+
+  private static final Logger logger = LoggerFactory.getLogger(Env.class);
+
+  // name of environment, cannot be null
+  private final String name;
+
+  // use to cache Env
+  private static final Map<String, Env> STRING_ENV_MAP = new ConcurrentHashMap<>();;
+
+  // default environments
+  public static final Env LOCAL = addEnv("LOCAL");
+  public static final Env DEV = addEnv("DEV");
+  public static final Env FWS = addEnv("FWS");
+  public static final Env FAT = addEnv("FAT");
+  public static final Env UAT = addEnv("UAT");
+  public static final Env LPT = addEnv("LPT");
+  public static final Env PRO = addEnv("PRO");
+  public static final Env TOOLS = addEnv("TOOLS");
+  public static final Env UNKNOWN = addEnv("UNKNOWN");
+
+  /**
+   * Cannot create by other
+   * @param name
+   */
+  private Env(String name) {
+    this.name = name;
+  }
+
+  /**
+   * a environment name exist or not
+   * @param name
+   * @return
+   */
+  public static boolean exist(String name) {
+    name = EnvUtils.getWellFormName(name);
+    return STRING_ENV_MAP.containsKey(name);
+  }
+
+  /**
+   * add an environment
+   * @param name
+   * @return
+   */
+  public static Env addEnv(String name) {
+    if (StringUtils.isBlank(name)) {
+      throw new RuntimeException("Cannot add a blank environment: " + "[" + name + "]");
+    }
+
+    name = EnvUtils.getWellFormName(name);
+    if(STRING_ENV_MAP.containsKey(name)) {
+      // has been existed
+      logger.info("{} has been existed.", name);
+    } else {
+      // not existed
+      logger.info("{} not existed. Now add it.", name);
+      STRING_ENV_MAP.put(name, new Env(name));
+    }
+    return STRING_ENV_MAP.get(name);
+  }
 
+  /**
+   * replace valueOf in enum
+   * But what would happened if environment not exist?
+   *
+   * @param name
+   * @throws IllegalArgumentException if this existed environment has no Env with the specified name
+   * @return
+   */
+  public static Env valueOf(String name) {
+    name = EnvUtils.getWellFormName(name);
```

Comment:
this logic is not the same as `transformEnv`, which will return Env.UNKNOWN instead of throwing an exception

Reply:
the method is for the backward compatibility with `enum`'s default method `valueOf`, which would throws IllegalArgumentException if this enum type has no constant with the specified name. 
Now `Env`'s type was changed to a `final class`, but there are still too many place use the old method `valueOf`, which will throws IllegalArgumentException. 
So I keep the exception. 
Should I delete the Exception?

Reply:
I swa many places which called `transformEnv` is not changed to call `valueOf`. Previously calling `transformEnv` with an invalid env will get Env.UNKNOWN instead of an exception.

Reply:
### about transformEnv
I keep the logic with master branch's `transformEnv` method. So the logic in `transformEnv` is same as before now.
### about valueOf
The method `valueOf` is auto generated by `Enum` type before, so for the same logic, I keep the exception in `valueOf`.